### ISSUE TITLE
snake_case sniff triggered by DOMNode properties

### ISSUE THREAD

**Issue by mundschenk-at** [WordPress/WordPress-Coding-Standards#574]
Third-party libraries like the DOM implementation included with PHP currently trigger the snake_case sniff. For example, the following code 

```
function get_ancestors( \DOMNode $node ) {
    $node = $node->parentNode 
}
```

results in `ERROR Object property "parentNode" is not in valid snake_case format`.

**Comment by jrfnl** [comment]
This is a typical use case for which I created PR #557. If you like to check it out and test it, it would be great to get your feedback on the PR.

**Comment by JDGrimes** [comment]
I've merged your PR, @jrfnl. I think that it would still be good to add any properties from libraries like DOM included with PHP to our list of defaults. PRs for that welcome. 😄

**Comment by jrfnl** [comment]
Thanks. I actually have a [list of PHP core class names](https://github.com/jrfnl/debug-bar-pretty-output/blob/master/class-debug-bar-list-php-classes.php) somewhere already (that class also triggers the error), but would need to check for property/method names. If we'd just look at PHP core, my guess would be that that would quickly become a very, very, very long whitelist...

In the mean time for project specific use, the option to whitelist variables which was now merged can tide people over.

**Comment by jrfnl** [comment]
I've had a look at this and am not sure this should actually be changed.
The thing is - how to determine from which class the property comes from ?

In the example given @mundschenk-at there's a type hint referencing a known native PHP class, but relating that type hint to the property being used in the line below is not actually that easy, let alone fool proof.
However, more often than not, there is no way to reliably determine if the non-snakecase property is from a known PHP class or from a custom class. And what about objects injected in the constructor and then used throughout a class ?

The only compromise I can currently think of would be to create a 'grey list' rather than a whitelist and if something is on the 'grey list' throw a warning instead of an error.
Possibly to be combined with a whitelist comment implementation ?

What do you think ?

**Comment by jrfnl** [comment]
Loosely related #623 - which takes care of preventing the ValidFunctionNames sniff from triggering if methods within an extended class do not comply with the snake case rule as often they can't as they are overloading CamelCase parent methods.

**Comment by mundschenk-at** [comment]
@jrfnl The suggested greylisting with a whitelist comment sounds like a good idea. (Side note: All sniffs (or at least almost all ;) should have specific whitelist comments. Having to use `@codingStandardsIgnoreStart/End` is dangerous.)

**Comment by dingo-d** [comment]
So this means that we should add all the methods that we might be using in the list like this:

~~~xml
  <rule ref="WordPress.NamingConventions.ValidVariableName">
    <properties>
        <property name="customVariablesWhitelist" value="preserveWhiteSpace" type="method" />
        <property name="customVariablesWhitelist" value="getElementById" type="method" />
    </properties>
  </rule>
~~~

?

**Comment by jrfnl** [comment]
> So this means that we should add all the methods that we might be using in the list like this:

@dingo-d I'm not sure I understand your question. The example xml you provided is mixing two things: you are trying to whitelist methods for the `ValidVariableName` sniff (which doesn't check method names) and doing it in an invalid way.

Could you provide some example PHP code and the error codes you are receiving for it so I can advise you better on how to handle it (and maybe open your own issue for this) ?

https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#mixed-case-property-name-exceptions

**Comment by dingo-d** [comment]
I need to scrape some links, and content from them for a client, so I am using `DOMDocument()` 

```php
$link = sanitize_text_field( wp_unslash( $_POST['link'] ) );

$dom = new DOMDocument();
$dom->preserveWhiteSpace = false;
$dom->loadHTMLFile( $link );
```

I got the `Object property "preserveWhiteSpace" is not in valid snake_case format` error. So I added 

```xml
  <rule ref="WordPress.NamingConventions.ValidVariableName">
    <properties>
        <property name="customVariablesWhitelist" value="preserveWhiteSpace" type="method" />
    </properties>
  </rule>
```

to my ruleset, and was wondering if I just need to add other such cases to this property, so that I can white list them.

**Comment by jrfnl** [comment]
@dingo-d Please have a look at the property documentation: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#mixed-case-property-name-exceptions

I don't know where you got the `type` from, but `method` is not a valid value.
The `type` should be `array` and the value a comma-separated list of the properties you want to whitelist.
Also note that the property name has changed in WPCS 0.11.0 to `customPropertiesWhitelist`.

This should work:
```xml
  <rule ref="WordPress.NamingConventions.ValidVariableName">
    <properties>
        <property name="customPropertiesWhitelist" type="array" value="preserveWhiteSpace"/>
    </properties>
  </rule>
```

If you want to add more properties:
```xml
  <rule ref="WordPress.NamingConventions.ValidVariableName">
    <properties>
        <property name="customPropertiesWhitelist" type="array" value="preserveWhiteSpace,anotherProperty,andYetAnother"/>
    </properties>
  </rule>
```

**Comment by dingo-d** [comment]
Oh, I see, thanks!!

**Comment by jrfnl** [comment]
Related to #1102 and #1623 which discuss adding the ability to scope whitelisted variables.

**Comment by jrfnl** [comment]
Closing as "mostly fixed":
* Property names in classes which extend other classes are now ignored by default.
* There is also the option to "allow list" specific names.

Tickets #1102 and #1623 which discuss scoping the allow list can be seen as a follow-up action.
